Userland/Libraries/LibTorrent/Bencode/Bencode.cpp

```diff
@@ -0,0 +1,232 @@
+        if (*end != 'e')
+            return {};
+
+        return adopt_own_if_nonnull(new BencodeNode { BencodeNodeType::Integer, value });
+    }
+    case 'd':
+    case 'l': {
+        auto maybe_node = decode_one(buffer, index);
+        if (maybe_node)
+            return maybe_node;
+        else
+            return {};
+    }
+    case '1':
+    case '2':
+    case '3':
+    case '4':
+    case '5':
+    case '6':
+    case '7':
+    case '8':
+    case '9': //must be a string
+    {
+        char* end;
+        i64 value = strtol(reinterpret_cast<char const*>(buffer.data()), &end, 10);
+        if (*end != ':' || value <= 0)
+            return {};
+        return adopt_own_if_nonnull(new BencodeNode { BencodeNodeType::String, String(end + 1, value) });
+    }
+    default:
+        return {};
+    }
+}
+
+OwnPtr<BencodeNode> decode(StringView const& view)
+{
+    return decode(view.bytes());
+}
+
+BencodeNode* BencodeNode::get_node_with_path(StringView const& path)
+{
+    VERIFY(!path.is_empty());
+    auto node = this;
+    auto parts = path.split_view('/');
+    if (parts.is_empty())
+        return node;
+
+    for (const auto& part : parts) {
+        switch (node->type()) {
+        case BencodeNodeType::Integer:
+        case BencodeNodeType::String:
+        case BencodeNodeType::KeyValuePair:
+            return nullptr;
+        case BencodeNodeType::List: {
+            auto maybe_index = part.to_uint();
+            if (maybe_index.has_value()) {
+                auto& list = node->get<NonnullOwnPtr<NonnullOwnPtrVector<BencodeNode>>>();
+                if (list->size() <= maybe_index.value())
+                    return nullptr;
+                else
+                    node = list->ptr_at(maybe_index.value()).ptr();
+            } else
+                return nullptr;
+        } break;
+        case BencodeNodeType::Dictionary: {
+            auto maybe_index = part.to_uint();
+            auto& list = node->get<NonnullOwnPtr<NonnullOwnPtrVector<BencodeNode>>>();
+            if (maybe_index.has_value()) {
+                if (list->size() <= maybe_index.value())
+                    return nullptr;
+                else
+                    node = list->ptr_at(maybe_index.value()).ptr();
+            } else {
+                bool found = false;
+                for (auto& kvnode : *list) {
+                    auto& pair = kvnode.get<NonnullOwnPtr<KeyValuePair>>();
+                    if (pair->key == part.substring_view(1, part.length() - 2)) {
+                        node = pair->value.ptr();
+                        found = true;
+                        break;
+                    }
+                }
+                if (!found)
+                    return nullptr;
+            }
+        } break;
+        }
+    }
+    return node;
+}
+BencodeNode const* BencodeNode::get_node_with_path(StringView const& path) const
+{
+    return const_cast<BencodeNode*>(this)->get_node_with_path(path);
```

Comment:
The way you wrote this, it could be that the non-`const` `get_node_with_path` somehow modifies the object, which it shouldn't. It of course cannot be `const` because it returns a non-`const` pointer into the object.

I suggest moving the actual logic into the `const` method, so we can be sure you don't accidentally modify the object in the method itself, and then just cast away the `const` on the returned pointer.